pkg/controllermanager/controller/credentialsbinding/reconciler.go

```diff
@@ -78,25 +82,43 @@ func (r *Reconciler) reconcile(ctx context.Context, credentialsBinding *security
 	}
 	kind := credential.GetObjectKind().GroupVersionKind().Kind
 
-	if !controllerutil.ContainsFinalizer(credential, gardencorev1beta1.ExternalGardenerName) {
+	if !controllerutil.ContainsFinalizer(credential, finalizerName) {
 		log.Info("Adding finalizer", kind, client.ObjectKeyFromObject(credential)) //nolint:logcheck
-		if err := controllerutils.AddFinalizers(ctx, r.Client, credential, gardencorev1beta1.ExternalGardenerName); err != nil {
+		if err := controllerutils.AddFinalizers(ctx, r.Client, credential, finalizerName); err != nil {
 			return fmt.Errorf("could not add finalizer to %s: %w", kind, err)
 		}
 	}
 
+	// TODO(dimityrmirchev): Remove the ExternalGardenerName finalizer handling in a future release
+	if controllerutil.ContainsFinalizer(credential, gardencorev1beta1.ExternalGardenerName) {
+		log.Info("Removing finalizer", kind, client.ObjectKeyFromObject(credential)) //nolint:logcheck
+		if err := controllerutils.RemoveFinalizers(ctx, r.Client, credential, gardencorev1beta1.ExternalGardenerName); err != nil {
+			return fmt.Errorf("could not remove finalizer from %s: %w", kind, err)
+		}
+	}
+
 	providerTypeLabelKey := v1beta1constants.LabelShootProviderPrefix + credentialsBinding.Provider.Type
 	labels := credential.GetLabels()
 
 	_, hasProviderKeyLabel := labels[providerTypeLabelKey]
 	_, hasCredentialsBindingRefLabel := labels[v1beta1constants.LabelCredentialsBindingReference]
-	if !hasProviderKeyLabel || !hasCredentialsBindingRefLabel {
+	secretNeedsLabel := !hasProviderKeyLabel && credential.GetObjectKind().GroupVersionKind().Kind == "Secret"
+	workloadIdentityNeedsLabelRemoval := hasProviderKeyLabel && credential.GetObjectKind().GroupVersionKind().Kind == "WorkloadIdentity"
+	if secretNeedsLabel || workloadIdentityNeedsLabelRemoval || !hasCredentialsBindingRefLabel {
 		patch := client.MergeFrom(credential.DeepCopyObject().(client.Object))
-		if !hasProviderKeyLabel {
+		if secretNeedsLabel {
 			credential.SetLabels(utils.MergeStringMaps(credential.GetLabels(), map[string]string{
 				providerTypeLabelKey: "true",
 			}))
+		} else if workloadIdentityNeedsLabelRemoval {
+			// TODO(dimityrmirchev): remove this handling in a future release
+			// workload identities do not need the provider shoot label
+			// they rely on the provider.extensions.gardener.cloud/<provider> label set by the extensionlabels admission plugin
```

Comment:
Is the label also not needed by any extension admission?

Reply:
I am not sure that I get the question. The label is used by the extensions https://github.com/gardener/gardener-extension-provider-gcp/blob/d4d3d43e7b3e6590e5f21c6f3d05ae0322ce3d83/pkg/admission/validator/webhook.go#L87